src/software/ai/navigator/path_planner/bang_bang_trajectory_2d_test.cpp

```diff
@@ -0,0 +1,114 @@
+#include "software/ai/navigator/path_planner/bang_bang_trajectory_2d.h"
+
+#include <gtest/gtest.h>
+
+#include <random>
+
+#include "software/test_util/test_util.h"
+
+class BangBangTrajectory2DTest : public testing::Test
+{
+   public:
+    // Use a fixed seed for the random number generator so that the tests are
+    // deterministic
+    BangBangTrajectory2DTest()
+        : rng(1010), pos_uniform_dist(-10, 10), vel_uniform_dist(-2, 2)
+    {
+    }
+
+   protected:
+    static const int NUM_SUB_POINTS   = 30;
+    static const int NUM_RANDOM_TESTS = 1000;
+
+    void verifyKinematicConstraints(double max_vel, double max_accel, double max_decel)
+    {
+        const double sub_point_length_sec = traj.getTotalTime() / NUM_SUB_POINTS;
+        const double max_accel_decel      = std::max(max_accel, std::abs(max_decel));
+
+        for (int i = 0; i <= NUM_SUB_POINTS; i++)
+        {
+            double t = i * sub_point_length_sec;
+
+            // Note that the velocity may slightly exceed the max velocity if one of
+            // the components gets a small fraction of the overall max velocity, and
+            // has an initial velocity which exceeds it. By the time it decelerates
+            // to something reasonable, the other component may be accelerating,
+            // resulting in the velocity vector exceeding the max.
+            Vector vel = traj.getVelocity(t);
+            ASSERT_LE(vel.length(), max_vel + 0.5)
+                << "Velocity constraint violated at t=" << t << " by "
+                << vel.length() - max_vel;
```

Comment:
so this fails in the first few iterations of the trajectory where we can't decelerate quickly enough such that the current velocity is still influenced by the large initial velocity?

Reply:
Im not sure if I fully understand your interpretations, but I think you're on the right track.

Basically, each 1D subtrajectory gets a component of the max velocity and max acceleration (e.g. max x velocity = 0.5m/s, max y velocity = 3.0m/s), if the initial velocity we start at is higher than this max velocity (e.g. the RNG selects 2.0m/s as the initial x velocity), then you could have a scenario where the y component quickly accelerates to the max velocity, while the x velocity is decelerating but still above the max x velocity. This leads to the overall velocity vector to go over the actual allowed 2D max velocity. Idk if I explained it well, but hope it helps.

Reply:
okay makes sense